src/main/java/com/example/Restaurant_Management_System_REST_API/service/CustomerService.java

```diff
@@ -0,0 +1,123 @@
+import com.example.Restaurant_Management_System_REST_API.repository.CustomerRepository;
+import com.example.Restaurant_Management_System_REST_API.service.generic.GenericBasicCrudOperations;
+import jakarta.validation.ConstraintViolation;
+import jakarta.validation.ConstraintViolationException;
+import jakarta.validation.Validator;
+import lombok.AllArgsConstructor;
+import org.hibernate.PropertyValueException;
+import org.modelmapper.ModelMapper;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.crypto.password.PasswordEncoder;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+@AllArgsConstructor
+public class CustomerService implements GenericBasicCrudOperations<CustomerDTOResponse, CustomerDTORequest, Long> {
+
+    private CustomerRepository customerRepository;
+    private AuthorityRepository authorityRepository;
+    private ModelMapper modelMapper;
+    private PasswordEncoder passwordEncoder;
+    private Validator validator;
+    
+    @Override
+    public CustomerDTOResponse create(CustomerDTORequest customerDTORequest) {
+
+        // First I Validate the CustomerDTORequest
+        Set<ConstraintViolation<CustomerDTORequest>> violations = validator.validate(customerDTORequest);
+        if (!violations.isEmpty()) {
+            throw new ConstraintViolationException(violations);
+        }
+
+        Customer customer = modelMapper.map(customerDTORequest, Customer.class);
+        customer.setPassword(passwordEncoder.encode(customerDTORequest.getPassword()));
+
+        if(customerDTORequest.getAuthorities() == null) {
+            throw new PropertyValueException("The authorities field in your request is null", "You tried to create a Customer",
+                    " But the missing field is: authorities.");
+        }
+
+        // Fetch the authorities from the database
+        // this is necessary because I don't want to have cascade in Customer, because I just want to have 4 roles in total
+        Set<Authority> authorities = customerDTORequest.getAuthorities().stream()
+                .map(authorityDTO -> {
+                    String authorityName = authorityDTO.getName();
+                    try {
+                        return authorityRepository.findByName(authorityName)
+                                .orElseThrow(() -> new NotFoundInDatabaseException(Authority.class));
+                    } catch (NotFoundInDatabaseException e) {
+                        throw new RuntimeException(e);
+                    }
+                })
+                .collect(Collectors.toSet());
+
+        customer.setAuthorities(authorities);
+        customerRepository.save(customer);
+
+        return modelMapper.map(customer, CustomerDTOResponse.class);
+    }
```

Comment:
There is a lot happening for this one `create` method. It'd be nice to split this method into separate, private methods or to another service for validation only. Right now this is not exactly following "**S**OLID" rule.

The method right now Is responsible for:
1) saving entity
2) mapping entity – which is send to ModelMapper (that's cool)
3) validating CustomerDTORequest
4) fetching authorities from db

The last two parts are pretty big and to follow good practices it's advisable to give them their own space instead of squishing it all in `create` method.

Reply:
done!